### PR TITLE
delete status deleted in status.go and modify related method in volume group.

### PR DESCRIPTION
**Pull request by lasshpanzer**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/api/db.go [hunk 1]
```diff
@@ -220,9 +222,9 @@ func DeleteVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) e
 
 //Just modify the state of the volume to be deleted in the DB, the real deletion in another thread
 func DeleteVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) error {
-	invalidStatus := []string{model.VolumeAvailable, model.VolumeError,
+	validStatus := []string{model.VolumeAvailable, model.VolumeError,
 		model.VolumeErrorDeleting, model.VolumeErrorExtending}
-	if !utils.Contained(in.Status, invalidStatus) {
+	if !utils.Contained(in.Status, validStatus) {
```
Comment (leonwanghui) [inline pkg/api/db.go]: Why this sentence is different from others?

File: pkg/api/db.go [hunk 2]
```diff
@@ -272,9 +274,9 @@ func EnableReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) error {
 }
 
 func DisableReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) error {
-	invalidStatus := []string{model.ReplicationCreating, model.ReplicationDeleting, model.ReplicationEnabling,
+	validStatus := []string{model.ReplicationCreating, model.ReplicationDeleting, model.ReplicationEnabling,
```
Comment (leonwanghui) [inline pkg/api/db.go]: It should be kept in `invalidStatus`.

File: pkg/api/db.go [hunk 3]
```diff
@@ -240,7 +242,7 @@ func DeleteReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) error {
 	invalidStatus := []string{model.ReplicationCreating, model.ReplicationDeleting, model.ReplicationEnabling,
 		model.ReplicationDisabling, model.ReplicationFailingOver, model.ReplicationFailingBack}
 
-	if utils.Contained(in.ReplicationStatus, invalidStatus) {
+	if utils.Contained(in.ReplicationStatus, validStatus) {
```
Comment (leonwanghui) [inline pkg/api/db.go]: I think you forgot to update this variable.

File: pkg/api/db.go [hunk 4]
```diff
@@ -257,7 +259,7 @@ func DeleteReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) error {
 func EnableReplicationDBEntry(ctx *c.Context, in *model.ReplicationSpec) error {
 	invalidStatus := []string{model.ReplicationCreating, model.ReplicationDeleting, model.ReplicationEnabling,
 		model.ReplicationDisabling, model.ReplicationFailingOver, model.ReplicationFailingBack}
-	if utils.Contained(in.ReplicationStatus, invalidStatus) {
+	if utils.Contained(in.ReplicationStatus, validStatus) {
```
Comment (leonwanghui) [inline pkg/api/db.go]: Same here

File: pkg/api/db.go (hunk 3 above)
Comment (lasshpanzer) [reply pkg/api/db.go]: OK，i am worried.

File: pkg/api/db.go (hunk 4 above)
Comment (lasshpanzer) [reply pkg/api/db.go]: OK，i am worried.

File: pkg/api/db.go (hunk 1 above)
Comment (lasshpanzer) [reply pkg/api/db.go]: My code logic is different from @xxwjj.

File: pkg/api/db.go [hunk 5]
```diff
@@ -205,7 +205,9 @@ func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (
 }
 
 func DeleteVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) error {
-	if in.Status != model.VolumeSnapAvailable {
+	validStatus := []string{model.VolumeSnapAvailable, model.VolumeSnapError,
+		model.VolumeSnapErrorDeleting, model.VolumeSnapDeleting}
+	if !utils.Contained(in.Status, validStatus) {
 		errMsg := "Only the volume snapshot with the status available can be deleted"
```
Comment (wisererik) [inline pkg/api/db.go]: please modify the errMsg

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -181,9 +195,21 @@ func (c *Controller) CreateVolume(ctx *c.Context, in *model.VolumeSpec, errchanV
 }
 
 func (c *Controller) DeleteVolume(ctx *c.Context, in *model.VolumeSpec, errchanvol chan error) {
+	// Profile id or pool id of the volume is not exist which means that volume creation failed before the driver
+	// was called , the volume entry should be delete from db directly.
+	if in.ProfileId == "" || in.PoolId == "" {
```
Comment (wisererik) [inline pkg/controller/controller.go]: please add log when delete volume from db failed.

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -115,15 +123,21 @@ func (c *Controller) CreateVolume(ctx *c.Context, in *model.VolumeSpec, errchanV
 
 	polInfo, err := c.selector.SelectSupportedPool(filterRequest)
 	if err != nil {
-		db.C.DeleteVolume(ctx, in.Id)
+		if errUpdate := db.C.UpdateStatus(ctx, in, model.VolumeError); errUpdate != nil {
```
Comment (leonwanghui) [inline pkg/controller/controller.go]: Did you roll back the code?

File: pkg/controller/controller.go (hunk 6 above)
Comment (lasshpanzer) [reply pkg/controller/controller.go]: ok

**Pull request merged**